### PR TITLE
Add benchmarks

### LINKED ISSUE #46: Benchmarking
As mentioned on the original GSoC proposals, this project is not complete until we validate that is well performant :)

For this, I've been cooking this repo: https://github.com/umccr/aws-benchmarks

As an example, using AWS S3 client libraries as subjects to test both CPU and I/O throughput on... but the idea is to apply those benches over here.

This task depends on #45 being finished.

### PR DESCRIPTION
**Pull request by CastilloDel**
Closes #46 

Before this is ready to get merge I need to add benchmarks for the htsget-search crate and I want to add running the benchmarks to the GitHub actions

### CODE REVIEW

**Review by brainstorm — CHANGES_REQUESTED** [review]
Fantastic work here, Daniel, well done!

Other than the couple of minor comments I left, I'd save one static benchmark that cleary shows the performance difference of the reference implementation against htsget-rs, as mentioned in Slack:

> Where does criterion-rs store the different results for the run(s)? Does it commit the .jsons back to the repository or save it in the GitHub Actions CI’s cache?
I’m thinking about having some sort of static plot in the README and then refer to the series of plots generated across releases, so that people can easily track the progress?

File: htsget-search/benches/benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+fn perform_query(query: Query) -> Result<(), HtsGetError> {
+  let htsget = HtsGetFromStorage::new(LocalStorage::new("../data", "localhost").unwrap());
+  htsget.search(query)?;
+  Ok(())
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+  let mut group = c.benchmark_group("Queries");
+  group
+    .sample_size(20)
```
[the remark this answers is not in the document]
Comment (CastilloDel) [reply htsget-search/benches/benchmark.rs:19]: Yes! Totally should've done that

File: htsget-http-actix/htsget-refserver-config.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
```
[the remark this answers is not in the document]
Comment (CastilloDel) [reply htsget-http-actix/htsget-refserver-config.json:1]: Seems reasonable

**Comment by brainstorm** [comment]
As for the commiting .htmls and jsons back to the repo... maybe https://github.com/boa-dev/criterion-compare-action is a better solution actually?

**Comment by CastilloDel** [comment]
It seems the action doesn't work :(
Something about an unsupported option
```
error: Unrecognized option: 'save-baseline'
error: bench failed
Error: The process 'cargo' failed with exit code 101
    at ExecState._setResult (/home/runner/work/_actions/jasonwilliams/criterion-compare-action/move_to_actions/dist/index.js:1:7880)
    at ExecState.CheckComplete (/home/runner/work/_actions/jasonwilliams/criterion-compare-action/move_to_actions/dist/index.js:1:7442)
    at ChildProcess.<anonymous> (/home/runner/work/_actions/jasonwilliams/criterion-compare-action/move_to_actions/dist/index.js:1:6406)
    at ChildProcess.emit (events.js:210:5)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
```
I'll put an issue

**Comment by CastilloDel** [comment]
I think I found a work-around! Although it means the criterion action will have to run twice

**Comment by CastilloDel** [comment]
Well, it failed again, but I think it is because it is trying to execute the benchmarks in the main branch, where it fails. Tomorrow I'll make a fork and try it 😥

**Comment by CastilloDel** [comment]
> `.sh`? :)

Yes, it is just a little script to set up the GA4HG server. It consists of only two commands, but one of them is rather long. I was using it myself, so I thought it could be convenient to add it to the project. Of course, I can just put those two commands in the instructions to set up the server!

**Comment by brainstorm** [comment]
Oh, sorry @CastilloDel, I just meant that the extension should be `.sh`:

Screen Shot 2021-08-18 at 11 08 26 am

**Comment by CastilloDel** [comment]
Ah! Yes!

**Comment by CastilloDel** [comment]
Fixed!

**Comment by mmalenic** [comment]
@brainstorm I think this might just be done now.

Thanks a lot @CastilloDel for all the code and guidance on how to structure the benchmarks.

Some recent changes:
* Refactor search and request benchmarks to use recent changes in htsget-rs.
* Some bug fixes in htsget-rs, such as byte ranges being inclusive for http headers.
* Add benchmark actions to main action.yml - runs only light benchmarks.

**Review by brainstorm — APPROVED** [review]
Amazing job, thanks @mmalenic! Nice forensics on the 1-off byte range bit, might have tricked you for a while :-!

Feel free to merge and move on to issues #85 and #75, we got this! :)

**Pull request merged**